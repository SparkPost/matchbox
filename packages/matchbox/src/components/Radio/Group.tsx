import React from 'react';
import PropTypes from 'prop-types';
import { Box } from '../Box';
import { Label } from '../Label';
import { ScreenReaderOnly } from '../ScreenReaderOnly';
import { OptionalLabel } from '../OptionalLabel';
import { Stack } from '../Stack';
import { createPropTypes } from '@styled-system/prop-types';
import styled from 'styled-components';
import { margin, MarginProps } from 'styled-system';
import { pick } from '@styled-system/props';

const StyledGroup = styled('fieldset')`
  border: none;
  padding: 0;
  ${margin}
`;

export type RadioGroupProps = {
  children: React.ReactNode;
  className?: string;
  'data-id'?: string;
  label: React.ReactNode;
  labelHidden?: boolean;
  optional?: boolean;
  required?: boolean;
} & MarginProps;

function Group(props: RadioGroupProps): JSX.Element {
  const {
    'data-id': dataId,
    children,
    className,
    label,
    labelHidden,
    required,
    optional,
    ...rest
  } = props;
  const systemProps = pick(rest);

  return (
    <StyledGroup {...systemProps} aria-required={required} className={className} data-id={dataId}>
      {label && (
        <Box width="100%">
          <Label as="legend" labelHidden={labelHidden}>
            <Box as="span" pr="200">
              {label}
              {required && <ScreenReaderOnly>Required</ScreenReaderOnly>}
            </Box>
            {optional && <OptionalLabel />}
          </Label>
        </Box>
      )}
      <Stack space="100">{children}</Stack>
    </StyledGroup>
  );
}

Group.displayName = 'Radio.Group';
export default Group;
