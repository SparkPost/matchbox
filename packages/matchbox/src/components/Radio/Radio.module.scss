@import '../../styles/config';

.Radio {
  position: relative;
  border: none;
  padding: 0;
  margin: 0 0 spacing() 0;
}

.Input {
  @include visually-hidden;

  &:hover:not(:disabled) {
    & ~ .Control .SvgUnchecked svg {
      fill: color(gray, 4);
    }
  }

  &:checked {
    & ~ .Control .SvgUnchecked {
      opacity: 0;
    }

    & ~ .Control .SvgChecked {
      opacity: 1;
    }
  }

  &:disabled  ~ .Control .SvgChecked,
  &:disabled  ~ .Control .SvgUnchecked {
    cursor: not-allowed;
    svg {
      fill: color(gray, 5);
    }
  }
}

.Control:hover {
  cursor: pointer;
}

.SvgUnchecked, .SvgChecked {
  position: absolute;
  top: 0;
  left: 0;

  &, & > svg {
    transition: 0.15s;
  }
}

.Control {
  display: block;
  position: relative;
  height: 21px;
  width: 21px;
}

.SvgUnchecked {
  opacity: 1;
  svg { fill: color(gray, 6) }
}

.SvgChecked {
  opacity: 0;
  svg { fill: color(blue) }
}

.Label {
  position: absolute;
  padding-left: 27px;
  padding-top: 3px;
  white-space: nowrap;
}

.Group {
  display: block;
  margin-bottom: spacing();

  & > * {
    margin-bottom: spacing(small);
  }

  &:last-child {
    margin: 0;
  }
}

.HelpText {
  padding-top: rem(5);
  font-size: font-size(300);
  line-height: line-height(300);
  color: color(gray, 4);
  padding-left: 27px;
}
