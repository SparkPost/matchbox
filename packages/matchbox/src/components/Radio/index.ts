export { default as Radio } from './Radio';
export * from './Radio';
export * from './Group';
