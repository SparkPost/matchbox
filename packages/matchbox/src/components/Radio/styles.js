import { tokens } from '@sparkpost/design-tokens';
import { Box } from '../Box';
import { visuallyHidden } from '../../styles/helpers';
import styled from 'styled-components';
import { margin } from 'styled-system';

export const Wrapper = styled('div')`
  ${margin}
  position: relative;
  border: none;
  padding: 0;
`;

export const StyledLabel = styled('label')`
  display: flex;
  align-items: center;
  margin: 0;
  padding: 0;
  white-space: nowrap;
  user-select: none;

  &:hover {
    cursor: ${props => (props.disabled ? 'not-allowed' : 'pointer')};
  }
`;

const commonCheckedStyles = `
  position: absolute;
  top: 0;
  left: 0;
  transition: ${tokens.motionDuration_fast} ${tokens.motionEase_in_out};
`;

export const StyledUnchecked = styled(Box)`
  ${commonCheckedStyles}
  opacity: 1;
  fill: ${props => (props.error ? tokens.color_red_700 : tokens.color_gray_700)};
`;

export const StyledChecked = styled(Box)`
  ${commonCheckedStyles}
  opacity: 0;
  fill: ${props => (props.error ? tokens.color_red_700 : tokens.color_blue_700)};
`;

export const StyledInput = styled('input')`
  ${visuallyHidden}

  &:hover:not(:disabled) {
    & ~ span ${StyledUnchecked} {
      fill: ${tokens.color_gray_800};
    }
  }

  &:checked {
    & ~ span ${StyledUnchecked} {
      opacity: 0;
    }

    & ~ span ${StyledChecked} {
      opacity: 1;
    }
  }

  &:disabled ~ span ${StyledChecked}, &:disabled ~ span ${StyledUnchecked} {
    cursor: not-allowed;
    fill: ${tokens.color_gray_600};
  }
`;
