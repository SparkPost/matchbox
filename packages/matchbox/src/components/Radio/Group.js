import React from 'react';
import PropTypes from 'prop-types';
import { Label } from '../Label';
import styles from './Radio.module.scss';

const Group = ({ children, label, required }) => (
  <div className={styles.Group}>
    {label && <Label className={styles.GroupLabel}>{label}{required && ' *'}</Label>}
    {children}
  </div>
);

Group.propTypes = {
  children: PropTypes.node.isRequired,
  label: PropTypes.node,
  required: PropTypes.bool
};

Group.displayName = 'Radio.Group';
export default Group;
