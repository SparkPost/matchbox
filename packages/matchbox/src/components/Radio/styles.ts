import { tokens } from '@sparkpost/design-tokens';
import { Box } from '../Box';
import { visuallyHidden } from '../../styles/helpers';
import styled from 'styled-components';
import { margin } from 'styled-system';

export const Wrapper = styled('div')`
  ${margin}
  position: relative;
  border: none;
  padding: 0;
`;

export const StyledLabel = styled('label')<{ $disabled?: boolean }>`
  display: flex;
  align-items: center;
  margin: 0;
  padding: 0;
  white-space: nowrap;
  user-select: none;

  &:hover {
    cursor: ${(props) => (props.$disabled ? 'not-allowed' : 'pointer')};
  }
`;

const commonCheckedStyles = `
  position: absolute;
  top: 0;
  left: 0;
  transition: opacity ${tokens.motionDuration_fast} ${tokens.motionEase_in_out}, fill ${tokens.motionDuration_fast} ${tokens.motionEase_in_out}, box-shadow ${tokens.motionDuration_fast} ${tokens.motionEase_in_out};
`;

export const StyledUnchecked = styled(Box)<{ $error?: string }>`
  ${commonCheckedStyles}
  opacity: 1;
  fill: ${(props) => (props.$error ? tokens.color_red_700 : tokens.color_gray_700)};
`;

export const StyledChecked = styled(Box)<{ $error?: string }>`
  ${commonCheckedStyles}
  opacity: 0;
  fill: ${(props) => (props.$error ? tokens.color_red_700 : tokens.color_blue_700)};
`;

export const StyledInput = styled('input')`
  ${visuallyHidden}

  & ~ span > * {
    border-radius: ${tokens.borderRadius_circle};
  }

  &:focus ~ span ${StyledChecked}, &:focus ~ span ${StyledUnchecked} {
    box-shadow: 0 0 0 0px ${tokens.color_white}, 0 0 0 2px ${tokens.color_blue_700};
  }

  &:checked {
    & ~ span ${StyledUnchecked} {
      opacity: 0;
    }

    & ~ span ${StyledChecked} {
      opacity: 1;
    }
  }

  &:hover:not(:disabled) {
    & ~ span ${StyledUnchecked} {
      fill: ${tokens.color_gray_800};
    }
  }

  &:disabled ~ span ${StyledChecked}, &:disabled ~ span ${StyledUnchecked} {
    cursor: not-allowed;
    fill: ${tokens.color_gray_600};
  }
`;
