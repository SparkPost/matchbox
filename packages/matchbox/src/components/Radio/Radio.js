import React from 'react';
import PropTypes from 'prop-types';
import { Label } from '../Label';
import { Error } from '../Error';
import { HelpText } from '../HelpText';
import { Box } from '../Box';
import useInputDescribedBy from '../../hooks/useInputDescribedBy';
import { RadioButtonChecked, RadioButtonUnchecked } from '@sparkpost/matchbox-icons';
import { pick } from '@styled-system/props';
import { omit } from '../../helpers/systemProps';
import { createPropTypes } from '@styled-system/prop-types';
import { margin } from 'styled-system';
import Group from './Group';
import { Wrapper, StyledLabel, StyledInput, StyledChecked, StyledUnchecked } from './styles';

function Radio(props) {
  const {
    id,
    name,
    checked,
    label,
    labelHidden,
    disabled,
    error,
    value,
    onChange,
    onFocus,
    onBlur,
    helpText,
    ...rest
  } = props;

  const componentProps = omit(rest, margin.propNames);
  const systemProps = pick(rest);

  const { describedBy, errorId, helpTextId } = useInputDescribedBy({
    id,
    hasHelpText: !!helpText,
    hasError: !!error,
  });

  return (
    <Wrapper {...systemProps}>
      <StyledLabel error={error} htmlFor={id} disabled={disabled}>
        <StyledInput
          id={id}
          name={name}
          value={value}
          checked={checked}
          disabled={disabled}
          onChange={onChange}
          onFocus={onFocus}
          onBlur={onBlur}
          type="radio"
          error={error}
          {...describedBy}
          {...componentProps}
        />
        <Box
          as="span"
          position="relative"
          display="block"
          alignItems="center"
          size="1rem" // Matches height of Label
        >
          <StyledUnchecked error={error} size="1rem" as={RadioButtonUnchecked} />
          <StyledChecked error={error} size="1rem" as={RadioButtonChecked} />
        </Box>
        <Box flex="1" pl="200">
          <Label
            as="span" // Outer wrapper already includes a label
            id={id}
            label={label}
            labelHidden={labelHidden}
            fontWeight="400"
            mb="0" // TODO Remove once margin 0 is baked into Label
          />
        </Box>
      </StyledLabel>
      {error && <Error id={errorId} error={error} />}
      {helpText && <HelpText id={helpTextId}>{helpText}</HelpText>}
    </Wrapper>
  );
}

Radio.displayName = 'Radio';

Radio.Group = Group;

Radio.propTypes = {
  id: PropTypes.string.isRequired,
  name: PropTypes.string,
  checked: PropTypes.bool,
  label: PropTypes.node,
  labelHidden: PropTypes.bool,
  disabled: PropTypes.bool,
  value: PropTypes.string,
  error: PropTypes.string,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  onBlur: PropTypes.func,
  helpText: PropTypes.node,
  ...createPropTypes(margin.propNames),
};

export default Radio;
