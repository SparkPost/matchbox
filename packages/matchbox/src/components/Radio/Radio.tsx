import React from 'react';
import { Label } from '../Label';
import { Error } from '../Error';
import { HelpText } from '../HelpText';
import { Box } from '../Box';
import useInputDescribedBy from '../../hooks/useInputDescribedBy';
import { RadioButtonChecked, RadioButtonUnchecked } from '@sparkpost/matchbox-icons';
import { pick } from '@styled-system/props';
import { omit } from '../../helpers/props';
import { margin, MarginProps } from 'styled-system';
import Group from './Group';
import { Wrapper, StyledLabel, StyledInput, StyledChecked, StyledUnchecked } from './styles';

export type RadioProps = {
  id: string;
  name?: string;
  checked?: boolean;
  label?: React.ReactNode;
  labelHidden?: boolean;
  disabled?: boolean;
  value?: string;
  error?: string;
  onChange?: React.ChangeEventHandler;
  onFocus?: React.FocusEventHandler;
  onBlur?: React.FocusEventHandler;
  helpText?: React.ReactNode;
} & MarginProps &
  Omit<React.ComponentPropsWithRef<'input'>, 'type'>;

const Radio = React.forwardRef<HTMLInputElement, RadioProps>(function Radio(props, userRef) {
  const {
    id,
    name,
    checked,
    label,
    labelHidden,
    disabled,
    error,
    value,
    onChange,
    onFocus,
    onBlur,
    helpText,
    ...rest
  } = props;

  const componentProps = omit(rest, margin.propNames);
  const systemProps = pick(rest);

  const { describedBy, errorId, helpTextId } = useInputDescribedBy({
    id,
    hasHelpText: !!helpText,
    hasError: !!error,
  });

  return (
    <Wrapper {...systemProps}>
      <StyledLabel htmlFor={id} $disabled={disabled}>
        <StyledInput
          aria-invalid={!!error}
          id={id}
          name={name}
          value={value}
          checked={checked}
          disabled={disabled}
          onChange={onChange}
          onFocus={onFocus}
          onBlur={onBlur}
          type="radio"
          ref={userRef}
          {...describedBy}
          {...componentProps}
        />
        <Box
          as="span"
          position="relative"
          display="block"
          alignItems="center"
          size="1rem" // Matches height of Label
        >
          <StyledUnchecked $error={error} size="1rem" as={RadioButtonUnchecked} />
          <StyledChecked $error={error} size="1rem" as={RadioButtonChecked} />
        </Box>
        <Box flex="1" pl="200">
          {label && (
            <Label
              as="span" // Outer wrapper already includes a label
              id={id}
              label={label}
              labelHidden={labelHidden}
              fontWeight="400"
              mb="0" // TODO Remove once margin 0 is baked into Label
            />
          )}
        </Box>
      </StyledLabel>
      {helpText && (
        <HelpText id={helpTextId} ml="500" mt="0">
          {helpText}
        </HelpText>
      )}
      {error && <Error id={errorId} error={error} ml="500" />}
    </Wrapper>
  );
}) as React.ForwardRefExoticComponent<RadioProps> & {
  Group: typeof Group;
};

Radio.displayName = 'Radio';

Radio.Group = Group;

export default Radio;
