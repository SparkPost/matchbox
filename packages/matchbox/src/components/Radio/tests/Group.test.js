import React from 'react';
import Radio from '../Radio';
import { shallow } from 'enzyme';

describe('Group', () => {

  it('renders children correctly', () => {
    expect(shallow(<Radio.Group>children</Radio.Group>)).toMatchSnapshot();
  });

  it('sets label and required', () => {
    expect(shallow(<Radio.Group label='label' required>children</Radio.Group>)).toMatchSnapshot();
  });
});
