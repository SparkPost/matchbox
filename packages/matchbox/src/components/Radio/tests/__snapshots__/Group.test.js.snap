// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Group renders children correctly 1`] = `
<div
  className="Group"
>
  children
</div>
`;

exports[`Group sets label and required 1`] = `
<div
  className="Group"
>
  <Label
    className="GroupLabel"
  >
    label
     *
  </Label>
  children
</div>
`;
